service_level_objectives_test.go

```diff
@@ -0,0 +1,213 @@
+		response, err := ioutil.ReadFile(fixturePath)
+		if err != nil {
+			t.Fatal(err)
+		}
+		w.Write(response)
+	}))
+
+	datadogClient := Client{
+		baseUrl:    ts.URL,
+		HttpClient: http.DefaultClient,
+	}
+	return ts, &datadogClient
+}
+
+func getMockSLO(id string) *ServiceLevelObjective {
+	return &ServiceLevelObjective{
+		ID:          &id,
+		Name:        sptr("Test SLO"),
+		Description: sptr("Test Description"),
+		Tags:        []string{"product:foo"},
+		Thresholds: []*ServiceLevelObjectiveThreshold{
+			{
+				TimeFrame: String("7d"),
+				SLO:       Float64(99),
+				Warning:   Float64(99.5),
+			},
+			{
+				TimeFrame: String("30d"),
+				SLO:       Float64(98),
+				Warning:   Float64(99),
+			},
+			{
+				TimeFrame: String("90d"),
+				SLO:       Float64(98),
+				Warning:   Float64(99),
+			},
+		},
+		Type:       &ServiceLevelObjectiveTypeMonitor,
+		MonitorIDs: []int{1},
+	}
+}
+
+func TestServiceLevelObjectiveIntegration(t *testing.T) {
+
+	t.Run("Create", func(t2 *testing.T) {
+		ts, c := testSLOGetMock(t2, "./tests/fixtures/service_level_objectives/create_response.json")
+		defer ts.Close()
+
+		slo := getMockSLO("")
+		created, err := c.CreateServiceLevelObjective(slo)
+		assert.NoError(t2, err)
+		assert.Equal(t2, "12345678901234567890123456789012", created.GetID())
+	})
+
+	t.Run("Update", func(t2 *testing.T) {
+		ts, c := testSLOGetMock(t2, "./tests/fixtures/service_level_objectives/update_response.json")
+		defer ts.Close()
+
+		slo := getMockSLO("12345678901234567890123456789012")
+		err := c.UpdateServiceLevelObjective(slo)
+		assert.NoError(t2, err)
+		assert.Equal(t2, "12345678901234567890123456789012", slo.GetID())
+		assert.Equal(t2, 1563283900, slo.GetModifiedAt())
+	})
+
+	t.Run("Delete", func(t2 *testing.T) {
+		ts, c := testSLOGetMock(t2, "./tests/fixtures/service_level_objectives/delete_response.json")
+		defer ts.Close()
+
+		slo := getMockSLO("12345678901234567890123456789012")
+		err := c.DeleteServiceLevelObjective(slo.GetID())
+		assert.NoError(t2, err)
+	})
+
+	t.Run("DeleteMany", func(t2 *testing.T) {
+		ts, c := testSLOGetMock(t2, "./tests/fixtures/service_level_objectives/delete_many_response.json")
+		defer ts.Close()
+
+		err := c.DeleteServiceLevelObjectives(
+			[]string{"12345678901234567890123456789012", "abcdefabcdefabcdefabcdefabcdefab"},
+		)
+		assert.NoError(t2, err)
+	})
+
+	t.Run("DeleteByTimeframe", func(t2 *testing.T) {
+		ts, c := testSLOGetMock(t2, "./tests/fixtures/service_level_objectives/delete_by_timeframe_response.json")
+		defer ts.Close()
+
+		resp, err := c.DeleteServiceLevelObjectiveTimeFrames(map[string][]string{
+			"12345678901234567890123456789012": {"7d"},
+			"abcdefabcdefabcdefabcdefabcdefab": {"30d", "90d"},
+		})
+		assert.NoError(t2, err)
+		assert.EqualValues(t2, resp.UpdatedIDs, []string{"12345678901234567890123456789012"})
+		assert.EqualValues(t2, resp.DeletedIDs, []string{"abcdefabcdefabcdefabcdefabcdefab"})
```

Comment:
`abcdefabcdefabcdefabcdefabcdefab` has two timeframes: 7d and 30d. Why does this request delete it?

Reply:
In the end it will have 0 specified time frames.

Reply:
Then I probably don't understand what the argument passed to `DeleteServiceLevelObjectiveTimeFrames` means. Intuitively, I'd expect it to delete `30d` and `90d` for SLO `abcdefabcdefabcdefabcdefabcdefab` and `7d` for SLO `12345678901234567890123456789012`. Does it do something different?

Reply:
it does do that exactly that. I'll add more comments here, but the idea is that: `12345678901234567890123456789012` has 2 (or more time frames) already specified, and `abcdefabcdefabcdefabcdefabcdefab` has *only* those 2 timeframes specified.

so in the case of `abcdefabcdefabcdefabcdefabcdefab` after deleting the 2 specified there will be 0 timeframes remaining, hence it's fully deleted

in the case of `12345678901234567890123456789012` it still has 1 timeframe remaining

Reply:
Well that's my point exactly. If you look at the fixture for `abcdefabcdefabcdefabcdefabcdefab`, you're going to see that it has `7d` and `30d`, so by the command above you'd only delete its `30d` timeframe, but leave the `7d` one. You probably meant to have `"abcdefabcdefabcdefabcdefabcdefab": {"7d", "30d"}` in the request.

Reply:
I'm not sure what fixture you're referring to here. The fixtures are only examples of the responses not the states prior. I've updated the comments in this test to relay the state prior to the test method.

Reply:
You created a fixture `tests/fixtures/service_level_objectives/create_response_metric.json` which contains an SLO with name `abcdefabcdefabcdefabcdefabcdefab`. This made it really confusing, as I thought this test case was also related to it and that didn't make sense. The added comment clears this up, so I have no further objections here.